README.md
hoauth v1.2.5
=============

* `hoauth` extension provides simple integration with social network authorization lib [Hybridauth](http://hybridauth.sourceforge.net/) in Yii. (facebook, google, twitter, vkontakte and much more).
* Automatically finds and supports `yii-user` module ([instruction for yii-user](https://github.com/SleepWalker/hoauth/wiki/%5Binstall%5D-hoauth-and-yii-user-extension)).
* supports prefixed tables

Requirements
------------
* Yii 1.1.9 or above. (I have tested it only in 1.1.13)

Available social networks
-------------------------

* OpenID
* Google
* Facebook
* Twitter
* Yahoo
* MySpace
* Windows Live
* LinkedIn
* Foursquare
* Vkontakte
* AOL

Additional social networks providers can be found at HybridAuth [website](http://hybridauth.sourceforge.net/download.html). And how to configure them [here](http://hybridauth.sourceforge.net/userguide.html) at the bottom of the page.

Available translations
----------------------
* English
* German
* Russian
* Spanish (Thanks to Komannder)

A little about how it's woks
----------------------------

The users of `yii-user` extension can skip this section, because `hoauth` will do all the stuff automatically.
This extension authenticates and if it's need creates new user. When user was registered "locally" (so he has login (email) and password), then he can also log in with it's social account (extension checks if user with provided email exists in db, when yes, the he will be logged in and it is no matter how had he registered earlier - locally or not). After the user logged in he will be redirected to `Yii::app()->user->returnUrl`.

This extension requires `UserIdentity` class, but doesn't use `authenticate()` method of `UserIdentity` class. Class constructor called with parameters `new UserIdentity($mail, null)` and than called `CWebUser::login()` method (while authentication work did for us social network). When social network didn't give us user's email, the **hoauth** will ask user for email, when email exists in our db, the password will be asked too. At the end we bind provided by social network unique user identifier to user id for future sign in. [Example of UserIdentity class](https://github.com/SleepWalker/hoauth/wiki/UserIdentity-class-example).

If you need to perform some access check for user, you can use `hoauthCheckAccess($user)` callback (simply create new method in controller where you added `HOAuthAction`). This method will be called with one input argument - *User model* of the user being authorized. This method should return integer values (`accessCode`) depending on the scenario needed:
* 0 - user shouldn't get access
* 1 - user may login
* 2 - user may login, but not now (e.g. the email should be verified and activated)
You can also not only return the `accessCode`, but also render the page with error or any information you need.

**NOTE:** This extension will also automatically create `user_oauth` table in your database. About it see "`UserOAuth` model" section.

Installation and Usage
----------------------

* [instruction for yii-user](https://github.com/SleepWalker/hoauth/wiki/%5Binstall%5D-hoauth-and-yii-user-extension)

**1\.** Make `hoauth` directory in your `extensions` directory (or in any other directory you want) and copy the content files there.
Directory structure example:
```php
/protected/
   extesions/
      hoauth/
         hybridauth/
         messages/
         models/
         views/
         widgets/
         .gitignore
         CHANGELOG
         DummyUserIdentity.php
         HOAuthAction.php
         HOAuthAdminAction.php
         MIT-LICENSE.txt
         README.md
         UPGRADE.md
```

**2\.** Edit your controller source code (eg. `SiteController` class with `actionLogin()` method) to add new actions:
```php
class SiteController extends Controller
{
	/**
	 * Declares class-based actions.
	 */
	public function actions()
	{
		return array(
      'oauth' => array(
        // the list of additional properties of this action is below
        'class'=>'ext.hoauth.HOAuthAction',
        // Yii alias for your user's model, or simply class name, when it already on yii's import path
        // default value of this property is: User
        'model' => 'User', 
        // map model attributes to attributes of user's social profile
        // model attribute => profile attribute
        // the list of avaible attributes is below
        'attributes' => array(
          'email' => 'email',
          'fname' => 'firstName',
          'lname' => 'lastName',
          'gender' => 'genderShort',
          'birthday' => 'birthDate',
          // you can also specify additional values, 
          // that will be applied to your model (eg. account activation status)
          'acc_status' => 1,
        ),
      ),
      // this is an admin action that will help you to configure HybridAuth 
      // (you must delete this action, when you'll be ready with configuration, or 
      // specify rules for admin role. User shouldn't have access to this action!)
      'oauthadmin' => array(
        'class'=>'ext.hoauth.HOAuthAdminAction',
      ),
		);
	}
}
```

**3\.** Add the `findByEmail` method to your user`s model class:
```php
  /**
   * Returns User model by its email
   * 
   * @param string $email 
   * @access public
   * @return User
   */
  public function findByEmail($email)
  {
    return self::model()->findByAttributes(array('email' => $email));
  }
```

**4\.** Visit your `oauthadmin` action (eg. http://yoursite.com/site/oauthadmin) to create the HybridAuth config. For your `HybridAuth Endpoint URL` use this: http://yoursite.com/site/oauth. After install you can leave `install.php` in your file system, while it's in Yii protected directory. But you must **remove** `oauthadmin` action, or make such rules, that give access only for admin users. Config file can be found at `application.config.hoauth`

**5\.** Add social login widget to your login page view (you can use `route` property, when you placing your widget not in the same module/controller as your `oauth` action):
```php
<?php $this->widget('ext.hoauth.widgets.HOAuth'); ?>
```
This widget can be switched to icon view using `onlyIcons` property. It may be also usefull the properties to adjust popup window size: `popupWidth` and `popupHeight`. See `HOAuth.php` file for details.

**Optional:**
**6\.** When you planning to use social networks like **Twitter**, that returns no email from user profile, you should declare `verifyPassword($password)` or `validatePassword($password)` method in `User` model, that should take the password (not hash) and return `true` if it is valid.
**7\.** You can also declare the `sendActivationMail()` method, that should mark the user account as inactive and send the mail for activation. This method, when it's exists will be used for social networks like **Twitter**, that give us no data about user's email (because we need to proof that user entered the right email).

Note
----
If you want to correctly display the facebook popup, you should add `"display" => "popup"` to the Facebook configuration array in `protected/config/hoauth.php`. E.g.:
```php
...
"Facebook" => array ( 
  "enabled" => true,
  "keys"    => array ( "id" => "PUT_YOURS_HERE", "secret" => "PUT_YOURS_HERE" ), 
  "scope"   => "email, user_about_me, user_birthday, user_hometown", // you can change the data, that will be asked from user
  "display" => "popup" // <- this one
)
...
```

Available social profile fields
-------------------------------

You can find them at HybridAuth [website](http://hybridauth.sourceforge.net/userguide/Profile_Data_User_Profile.html).
And here is some additional fields, that I needed in my project, you can use them too:
* `birthDate` - The full date of birthday (eg. 1991-09-03)
* `genderShort` - short representation of gender (eg. 'm', 'f')

Additional properties for `HOAuthAction`
----------------------------------------
* `useYiiUser` - enables support for `yii-user` (default: false). `hoauth` will find `yii-user` module automatically, so you can leave this property as default. You may also leave `attributes` and `model` properties as default.
* `enabled` - defines whether the ouath functionality is active. Useful for example for CMS, where user can enable or disable oauth functionality in control panel. (default: true)
* `scenario` - scenario name for the $model (optional)
* `loginAction` - name of a local login action (should be in the same controller as `oauth` action). (default: 'actionLogin')
* `duration` - 'remember me' duration in ms. (default: 2592000 //30days)
* `usernameAttribute` - you can specify username attribute, when it must be unique (like in `yii-user` extension), that hoauth will try to validate it's uniqueness.
* `alwaysCheckPass` - flag to control password checking for the scenario, when when social network returned email of existing local account. If set to `false` user will be automatically logged in without confirming account with password. (default: `true`)

Available Callbacks
-------------------
To make you able to customize the behavior of `hoauth`, it is provide some useful callbacks. Here the list of them:
* `Controller::hoauthCheckAccess($user)`
* `Controller::hoauthAfterLogin($user)`
* `User::findByEmail($email`
* `User::verifyPassword($password)` or `User::validatePassword($password)`
* `User::sendActivationMail())`

[More about callbacks](https://github.com/SleepWalker/hoauth/wiki/Callbacks)

`UserOAuth` model
-----------------

`UserOAuth` model used to bind social services to user's account and to store session with social network profile. If you want to use this data (user profile) later, please use `UserOAuth::getProfile()` method:
```php
$userOAuths = UserOAuth::model()->findUser(5); // find all authorizations from user with id=5
foreach($userOAuths as $userOAuth)
{
  $profile = $userOAuth->profile;
  echo "Your email is {$profile->email} and social network - {$userOAuth->provider}<br />";
}
```
or
```php
$userOAuth = UserOAuth::model()->findUser(5, "Google"); // find all authorizations from user with id=5
$profile = $userOAuth->profile;
echo "Your email is {$profile->email} and social network - {$userOAuth->provider}<br />";
```
You can also use `UserOAuth::profileCache` property to access cached copy of the profile, without making any request to the social network.
About how to use HybridAuth object you can read [here](http://hybridauth.sourceforge.net/userguide.html).

Sources
-------

* [HybridAuth] (http://hybridauth.sourceforge.net)
* [Zocial CSS3 Buttons] (https://github.com/samcollins/css-social-buttons/)
* [Project page on Yii] (http://yiiframework.com/extension/hoauth/)
* [instruction for yii-user](https://github.com/SleepWalker/hoauth/wiki/%5Binstall%5D-hoauth-and-yii-user-extension)


CHANGELOG
Version 1.2.5
-------------
- added an option to display only icons in social login widget (HOAuth::onlyIcons)
- Authentication through pop-up window
- Added support of CWebUser::returnUrl (for popup window only if it has not default value)
- Enh#45: Added composer support
- Small refactoring

Version 1.2.4
------------
- updated hybridauth lib
- support of older version of yii-user
- moved to MIT license

Version 1.2.3
-----------
- yii-user: we display `username` field only if it is required by the `User` model, so you can disable `username` field in `yii-user`.
- fixed bug when after saving profileCache returned in serialized form
- removed some NOTICE erros and fixed one stupid big with email sending callback (thanks to Komannder)
- fixed bug with imposibility of login with twitter (and other similar SN) despite successful email activation
- added two callbacks: hoauthCheckAccess() and hoauthAfterLogin(). Details at https://github.com/SleepWalker/hoauth/wiki/Callbacks.
- added Spanish, German and Russian translations (thanks to Komannder and me)
- Security improvement: if user try to login with email of existing local account, that is not bond to current social network, he will be asked for password to confirm binding.
- added support of validatePassword() callback - alias for verifyPassword()
- added widget HOAuthActive - display login buttons of social networks bond to current user (Charles)

Version 1.2.2
-------------
- Enh#6: support of prefixed table names
- Enh#7: added ability to setup alias (by default is application.config.hoauth) of HybridAuth config file in yii config (Yii::app()->params['hoauth']['configAlias'])
- Enh#8: widget to display social networks that user bond to
- Support of `yii-user` version, when Profile::regMode isn't static property
- Fixed issue when password field apeared, when it should not do so during social network signup
- Support of sending activation email by `yii-user` module
- Added support of `yii-user` banned and not activated account status
- Ability to register a new account with the same SN, when in db still exists SN relation to account that was deleted
- HOAuthWidget moved to widgets directory (see UPGRADE.md)

Version 1.2.1
-------------
- It was decieded to move `HybridAuth` config file to the yii's config diretory and rename to `hoauth.php`. Extension will try to do it automatically, when config diretory is not writable, it will run with config file from old directory, but old directory has the deprecated status.
- fixed bug with to long username, when registering user for `yii-user` extension.
- Support of login from social networks, that returning no email (also added `HUserInfoForm` class).
- updates in installation instructions

Version 1.2
-----------

- HOAuthWiget property `$controllerId` replaced by `$route` and now you can specify route e.g. module/controller
- Modification of **HybridAuth** install script to generating `Endpoint URL` properly.
- Renaming of `user_oauth` table columns: `name` -> `provider`, `value` -> `identifier` (model will automatically update schema)
- New features in `UserOAuth.php` model.
- Support for `yii-user` extension


HOAuthAction.php
<?php
/**
 * HOAuthAction - this the main class in hoauth extension.
 * 
 * @uses CAction
 * @version 1.2.5
 * @copyright Copyright &copy; 2013 Sviatoslav Danylenko
 * @author Sviatoslav Danylenko <dev@udf.su> 
 * @license MIT ({@link http://opensource.org/licenses/MIT})
 * @link https://github.com/SleepWalker/hoauth
 */

/**
 * HOAuthAction provides simple integration with social network authorization lib Hybridauth in Yii.
 *
 * HOAuthAction requires, that your user model implements findByEmail() method, that should return user model by its email.
 *
 * Avaible social networks: 
 *    OpenID, Google, Facebook, Twitter, Yahoo, MySpace, Windows Live, LinkedIn, Foursquare, AOL
 * Additional social networks can be found at: {@link http://hybridauth.sourceforge.net/download.html}
 *
 * Social Auth widget:
 *    <?php $this->widget('ext.hoauth.HOAuthWidget', array(
 *      'controllerId' => 'site', // id of controller where is your oauth action (default: site)
 *    )); ?>
 * uses a little modified Zocial CSS3 buttons: {@link https://github.com/samcollins/css-social-buttons/}
 */


Yii::setPathOfAlias('hoauth', dirname(__FILE__));

class HOAuthAction extends CAction
{
	/**
	 * @var boolean $enabled defines whether the ouath functionality is active. Useful for example for CMS, where user can enable or disable oauth functionality in control panel
	 */
	public $enabled = true;

	/**
	 * @var string $model yii alias for user model (or class name, when this model class exists in import path)
	 */
	public $model = 'User';

	/**
	 * @var array $attributes attributes synchronization array (user model attribute => oauth attribute). List of available profile attributes you can see at {@link http://hybridauth.sourceforge.net/userguide/Profile_Data_User_Profile.html "HybridAuth's Documentation"}.
	 *
	 * Additional attributes:
	 *    birthDate - The full date of birthday, eg. 1991-09-03
	 *    genderShort - short representation of gender, eg. 'm', 'f'
	 *
	 * You can also set attributes, that you need to save in model too, eg.:
	 *    'attributes' => array(
	 *      'is_active' => 1,
	 *      'date_joined' => new CDbExpression('NOW()'),
	 *    ),
	 *
	 * @see HOAuthAction::$avaibleAtts
	 */
	public $attributes = array();

	/**
	 * @var string $scenario scenario name for the $model (optional)
	 */
	public $scenario = 'insert';

	/**
	 * @var string $loginAction name of a local login action
	 */
	public $loginAction = 'actionLogin';

	/**
	 * @var integer $duration how long the script will remember the user
	 */
	public $duration = 2592000; // 30 days

	/**
	 * @var boolean $useYiiUser enables support of Yii user module
	 */
	public static $useYiiUser;

	/**
	 * @var boolean $alwaysCheckPass flag to control password checking for the scenario, 
	 *      when when social network returned email of existing local account. If set to
	 *      `false` user will be automatically logged in without confirming account with password
	 */
	public $alwaysCheckPass = true;

	/**
	 * @var string $userIdentityClass UserIdentity class that will be used to log user in.
	 */
	public $userIdentityClass = 'UserIdentity';

	/**
	 * @var string $usernameAttribute you can specify the username attribute, when user must fill it
	 */
	public $usernameAttribute = false;

	/**
	 * @var boolean $useUserReturnUrl enable/disable usage of CWebUser::returnUrl
	 */
	public $useUserReturnUrl = true;

	/**
	 * @var string|array $oauthReturnUrl specific for oauth return url
	 */
	public $oauthReturnUrl = false;

	/**
	 * @var string $returnUrlStateId the id that should be used to store state for hoauth return url
	 */
	public static $oauthReturnUrlStateId = '__oauthReturnUrl';

	/**
	 * @var string $_emailAttribute
	 */
	protected $_emailAttribute = false;

	/**
	 * @var array $avaibleAtts Hybridauth attributes that support by this script (this a list of all available attributes in HybridAuth 2.0.11) + additional attributes (see $attributes property)
	 */
	protected $_avaibleAtts = array('identifier', 'profileURL', 'webSiteURL', 'photoURL', 'displayName', 'description', 'firstName', 'lastName', 'gender', 'language', 'age', 'birthDay', 'birthMonth', 'birthYear', 'email', 'emailVerified', 'phone', 'address', 'country', 'region', 'city', 'zip', 'birthDate', 'genderShort');

	/**
	 * @var UserOAuth the model to handle the work during the request
	 */
	protected $_oauth;

	public function run()
	{		
		if($this->enabled) {
			$this->setUp();

			if(isset($_GET['provider'])) {
				// after oauth — working with user model and his data from SN
				Yii::import('hoauth.models.*');
				$this->oAuth($_GET['provider']);
			} else {
				// Handling OAuth (redirects, tokens etc.)
				$path = dirname(__FILE__);
				require($path.'/hybridauth/index.php');
				Yii::app()->end();
			}
		}

		Yii::app()->controller->{$this->loginAction}();
	}

	/**
	 * Initiates authorization with specified $provider and 
	 * then authenticates the user, when all goes fine
	 * 
	 * @param mixed $provider provider name for HybridAuth
	 * @access protected
	 * @return void
	 */
	protected function oAuth($provider)
	{
		try{
			// trying to authenticate user via social network
			$oAuth = UserOAuth::model()->authenticate($provider);
			$this->setOauth($oAuth);

			$accessCode = $this->getAccessCode();

			if($accessCode === 1) {
				// the authentication was successfull. closing auth window
				?>
				<script>
					var returnUrl = <?php echo $this->useUserReturnUrl ? CJavaScript::encode(Yii::app()->user->getReturnUrl(false)) : false; ?>;
					if(window.opener) {
						if(returnUrl) {
							window.opener.location.href = returnUrl;
						} else {
							window.opener.location.reload();
						}

						window.close();
					}else{
						window.location.href = returnUrl ? returnUrl : '/';
					}
				</script>
				<?php
				Yii::app()->end();
			}
		} catch(Exception $e) {
			$this->handleError($e);
		}
		?>
		<script>
			window.close();
		</script>
		<?php
	}

	/**
	 * @param  UserOAuth $oAuth
	 * @return integer        access code
	 *     $accessCode == 0 - user shouldn't get access
	 *     $accessCode == 1 - user may login
	 *     $accessCode == 2 - user may login, but not now (e.g. the email should be verified and activated)
	 */
	protected function getAccessCode()
	{
		$oAuth = $this->oAuth;
		// If we already have a user logged in, associate the authenticated 
		// provider with the logged-in user
		if(!Yii::app()->user->isGuest) {
			$accessCode = 1;
			$oAuth->bindTo(Yii::app()->user->id);
		} else {
			list($user, $isNewUser) = $this->prepareGuestUser();

			// checking if current user is not banned or anything else
			$accessCode = 1;
			if(method_exists($this->controller, 'hoauthCheckAccess')) {
				$accessCode = $this->controller->hoauthCheckAccess($user);
			} elseif($this->useYiiUser) {
				$accessCode = $this->yiiUserCheckAccess($user);
			}

			// sign user in
			switch($accessCode) {
				case 1:
					if(!$oAuth->bindTo($user->primaryKey)) {
						throw new Exception("Error, while binding user to provider:\n\n" . var_export($oAuth->errors, true));
					}

					$identity = $this->useYiiUser
						? new DummyUserIdentity($user->primaryKey, $user->email)
						: new $this->userIdentityClass($user->email, null);

					if(!Yii::app()->user->login($identity,$this->duration)) {
						throw new Exception("Can't sign in, something wrong with UserIdentity class.");
					}

					// user was successfully logged in
					// firing callback
					if(method_exists($this->controller, 'hoauthAfterLogin')) {
						$this->controller->hoauthAfterLogin($user, $isNewUser);
					}
					break;
					
				case 2: // stopping script to let checkAccess() function render new content
					Yii::app()->end();
					break;

				default:
					throw new Exception("Something wrong. You can not log in.");
			}
		}

		return $accessCode;
	}

	/**
	 * Prepares the model of the guest user,
	 * registers new user, if needed
	 * 
	 * @return array    of two elements:
	 *                  - user model
	 *                  - whether it is the new user
	 */
	protected function prepareGuestUser()
	{
		$isNewUser = false;
		$userProfile = $this->oAuth->profile;

		if($this->oAuth->isBond) {
			// this social network account is bond to existing local account
			Yii::log("Logged in with existing link with '$provider' provider", CLogger::LEVEL_INFO, 'hoauth.'.__CLASS__);
			$user = $this->userModel->findByPk($this->oAuth->user_id);

			return array($user, $isNewUser);
		} elseif(!empty($userProfile->emailVerified)) { // SN returned user email
			// checking whether we already have a user with specified email
			if($this->useYiiUser) {
				$user = $this->userModel->findByAttributes(array('email' => $userProfile->emailVerified));
			} else {
				$user = $this->userModel->findByEmail($userProfile->emailVerified);
			}
		}

		if(!isset($user)) {
			// registering a new user
			$user = new $this->model($this->scenario);
			$isNewUser = true;
		}

		if($this->alwaysCheckPass || $isNewUser) {
			if(method_exists($this->controller, 'hoauthProcessUser')) {
				$user = $this->controller->hoauthProcessUser($user, $isNewUser);
			} else {
				$user = $this->processUser($user, $userProfile);
			}
		}

		return array($user, $isNewUser);
	}
	
	/**
	 * Handles two cases:
	 * 	- new user (registers new user, collects username and email if needed)
	 * 	- the case, when social network returned email of existed user and
	 * 	  he is logging the first time with this SN. We will ask him for password to validate, that it is realy his account
	 *
	 * @param CActiveRecord $user current user model
	 * @param object $userProfile social network's user profile object
	 * @access protected
	 */
	protected function processUser($user, $userProfile)
	{
		if($this->useYiiUser) {
			$profile = new Profile();
			// enabling register mode
			// old versions of yii
			$profile->regMode = true;
			// new version, when regMode is static property
			$prop = new ReflectionProperty('Profile', 'regMode');
			if($prop->isStatic())
				Profile::$regMode = true;
		}

		if($user->isNewRecord) { // the old user should manage his profile by himself
			$this->populateModel($user, $userProfile);
		}

		// checking if we have all data to register new user
		// if SN returned email of existed user, check that he knows the password
		$user = $this->checkUserData($user);

		// the model won't be new, if user provided email and password of existing account
		if($user->isNewRecord) {
			if(!$user->save()) {
				throw new Exception("Error, while saving {$this->model} model:\n\n" . var_export($user->errors, true));
			}

			if($this->useYiiUser) {
				$profile->user_id = $user->primaryKey;
				if($profile->hasAttribute('firstname')) {
					// we have an older yii-user version or the db dump from `data` directory instead of migration was used
					$profile->firstname = $userProfile->firstName;
					$profile->lastname = $userProfile->lastName;
				} else {
					$profile->first_name = $userProfile->firstName;
					$profile->last_name = $userProfile->lastName;
				}

				if(!$profile->save()) {
					throw new Exception("Error, while saving " . get_class($profile) . " model:\n\n" . var_export($profile->errors, true));
				}
			}

			// trying to send activation email
			$this->sendActivationEmail($user);
		}

		return $user;
	}

	/**
	 * Uses HUserInfoForm to check if we have all data, that we need from new user
	 * displays the form to get the required, but not specified user data
	 * 
	 * @param  CActiveRecord $user user model
	 * @return CActiveRecord user model with correct data
	 */
	protected function checkUserData($user)
	{
		// trying to fill email and username fields
		// NOTE: we display `username` field in our form only if it is required by the model
		if($this->usernameAttribute && !$user->isAttributeRequired($this->usernameAttribute)) {
			$this->usernameAttribute = false;
		}

		$form = new HUserInfoForm($user, $this->_emailAttribute, $this->usernameAttribute);

		if(!$form->validateUser()) {
			// We need to request some info from user
			$this->controller->render('hoauth.views.form', array(
				'form' => $form,
				));
			Yii::app()->end();
		}

		// updating attributes in $user model (if needed)
		$user = $form->validUserModel;

		return $user;
	}

	/**
	 * Sends email activation email, when it is needed	
	 * 
	 * @param CActiveRecord $user current user model
	 * @access protected
	 * @return void
	 */
	protected function sendActivationEmail($user)
	{
		if($this->useYiiUser) {
			// why not to put this code not in controller, but in the User model of `yii-user` module?
			// for now I can only copy-paste this code from controller...
			if (Yii::app()->getModule('user')->sendActivationMail) {
				$activation_url = Yii::app()->createAbsoluteUrl('/user/activation/activation',array("activkey" => $user->activkey, "email" => $user->email));
				UserModule::sendMail($user->email,UserModule::t("You registered on {site_name}",array('{site_name}'=>Yii::app()->name)),UserModule::t("To activate your account, please go to {activation_url}",array('{activation_url}'=>$activation_url)));
			}
		} else {
			if(method_exists($user, 'sendActivationEmail')) {
				$user->sendActivationEmail();
			} elseif(method_exists($user, 'sendActivationMail')) { // TODO: delete in future
				$user->sendActivationMail();
			}
		}
	}

	/**
	 * Populates User model with data from social network profile
	 * 
	 * @param CActiveRecord $user users model
	 * @param mixed $profile HybridAuth user profile object
	 * @access protected
	 */
	protected function populateModel($user, $profile)
	{
		foreach($this->attributes as $attribute => $pAtt) {
			if(in_array($pAtt, $this->_avaibleAtts)) {
				switch($pAtt) {
					case 'genderShort':
						$gender = array('female'=>'f','male'=>'m');
						$att = $gender[$profile->gender];
						break;
					case 'birthDate':
						$att = $profile->birthYear 
						? sprintf("%04d-%02d-%02d", $profile->birthYear, $profile->birthMonth, $profile->birthDay)
						: null;
						break;
					case 'email':
						$att = $profile->emailVerified;
						break;
					default:
						$att = $profile->$pAtt;
				}
				if(!empty($att)) {
					$user->$attribute = $att;
				}
			} else {
				$user->$attribute = $pAtt;
			}
		}
	}

	/**
	 * Checks whether the $user can be logged in
	 *
	 * @param CActiveRecord $user current `yii-user` user's model
	 * @param boolean $render flag that enables rendering
	 */
	protected function yiiUserCheckAccess($user, $render = true)
	{
		if($user->status==0&&Yii::app()->getModule('user')->loginNotActiv==false) {
			$error = UserIdentity::ERROR_STATUS_NOTACTIV;
			$return = 2;
		} elseif($user->status==-1) {
			$error = UserIdentity::ERROR_STATUS_BAN;
			$return = 0;
		} else {
			$error = UserIdentity::ERROR_NONE;
			$return = 1;
		}

		if($error && $render) {
			$this->controller->render('hoauth.views.yiiUserError', array(
				'errorCode' => $error,
				'user' => $user,
				));
		}

		return $return;
	}

	/**
	 * Checks if the action was properly setup and ready to run
	 * @throws Exception If improperly setted up
	 */
	protected function setUp()
	{
		if(!is_array($this->attributes)) {
			$this->attributes = array();
		}

		if(!isset(self::$useYiiUser)) {
			self::$useYiiUser = file_exists(Yii::getPathOfAlias('application.modules.user.components') . '/UWrelBelongsTo.php');
		}

		// checking if we have `yii-user` module (I think that `UWrelBelongsTo` is unique class name from `yii-user`)
		if(self::$useYiiUser === true) {
			// setting up yii-user's user model
			Yii::import('application.modules.user.models.*');
			Yii::import('hoauth.DummyUserIdentity');

			// preparing attributes array for `yii-user` module
			$this->attributes = CMap::mergeArray(array(
				'email' => 'email',
				'username' => 'displayName',
				'status' => User::STATUS_ACTIVE,
				), $this->attributes);

			$this->usernameAttribute = 'username';
			$this->_emailAttribute = 'email';
		} else {
			if(!in_array('email', $this->attributes)) {
				throw new CException("You forgot to bind 'email' field in " . __CLASS__ . "::attributes property.");
			}

			Yii::import($this->model, true);
			$this->model = substr($this->model, strrpos($this->model, '.'));

			$this->_emailAttribute = array_search('email', $this->attributes);
		}

		if(empty($this->model) || !class_exists($this->model)) {
			throw new CException('You should specify the User model to work with');
		}

		if(!method_exists($this->model, 'findByEmail') && !self::$useYiiUser) {
			throw new Exception("Model '{$this->model}' must implement the 'findByEmail' method");
		}
	}

	/**
	 * @return CActiveRecord user model depending on existence of yii-user
	 */
	public function getUserModel()
	{
		if($this->useYiiUser) {
			$user = User::model();
		} else {
			$user = call_user_func(array($this->model, 'model'));
		}

		return $user;
	}

	public function getUseYiiUser()
	{
		return self::$useYiiUser;
	}

	public function setUseYiiUser($value)
	{
		self::$useYiiUser = $value;
	}

	public function getEmailAttribute()
	{
		return $this->_emailAttribute;
	}

	/**
	 * Sets the UserOAuth model to work with
	 * @param UserOAuth $value model
	 */
	public function setOauth(UserOAuth $value)
	{
		if(!$this->_oauth) {
			$this->_oauth = $value;
		}
	}

	/**
	 * Sets the UserOAuth model to work with
	 * @param UserOAuth $value model
	 */
	public function getOauth()
	{
		return $this->_oauth;
	}

	public static function t($message,$params=array(),$source=null,$language=null)
	{
		return Yii::t('HOAuthAction.root', $message,$params,$source,$language);
	}

	/**
	 * Handles, log or displays errors
	 * @param  Exception $e
	 */
	protected function handleError(Exception $e)
	{
		$error = "";

		// Display the received error
		switch($e->getCode()) { 
			case 0 : $error = "Unspecified error."; throw $e; break;
			case 1 : $error = "Hybriauth configuration error."; break;
			case 2 : $error = "Provider not properly configured."; break;
			case 3 : $error = "Unknown or disabled provider."; break;
			case 4 : $error = "Missing provider application credentials."; break;
			case 5 : $error = "Authentication failed. The user has canceled the authentication or the provider refused the connection."; break;
			case 6 : $error = "User profile request failed. Most likely the user is not connected to the provider and he should to authenticate again."; 
				@$this->oAuth->logout(); 
			break;
			case 7 : $error = "User not connected to the provider."; 
				@$this->logout(); 
			break;
			case 8 : $error = "Provider does not support this feature.";	break;
		}

		$error .= "\n\n<br /><br /><b>Original error message:</b> " . $e->getMessage(); 
		Yii::log(strip_tags($error), CLogger::LEVEL_INFO, 'hoauth.'.__CLASS__);
		if(YII_DEBUG)
			throw $e;
	}
}
